Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -0,0 +1,447 @@
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/master/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Hand written storage class for Google_Protobuf_Any to support on demand
+/// transforms between the formats.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+internal func buildTypeURL(forMessage message: Message, typePrefix: String) -> String {
+  var url = typePrefix
+  if typePrefix.isEmpty || typePrefix.characters.last != "/" {
+    url += "/"
+  }
+  return url + typeName(fromMessage: message)
+}
+
+fileprivate func typeName(fromMessage message: Message) -> String {
+  let messageType = type(of: message)
+  return messageType.protoMessageName
+}
+
+fileprivate func typeName(fromURL s: String) -> String {
+    var typeStart = s.startIndex
+    var i = typeStart
+    while i < s.endIndex {
+        let c = s[i]
+        i = s.index(after: i)
+        if c == "/" {
+            typeStart = i
+        }
+    }
+
+    return s[typeStart..<s.endIndex]
+}
+
+internal class AnyMessageStorage {
+  var _typeURL: String = ""
+
+  // Computed to do on demand work.
+  var _value: Data {
+    get {
+      if let value = _valueData {
+        return value
+      }
+
+      if let message = _message {
+        do {
+          return try message.serializedData(partial: true)
+        } catch {
+          return Data()
+        }
+      }
+
+      if let contentJSON = _contentJSON, !_typeURL.isEmpty {
+        let encodedTypeName = typeName(fromURL: _typeURL)
+        if let messageType = Google_Protobuf_Any.lookupMessageType(forMessageName: encodedTypeName) {
+          do {
+            // Hack, make an any to use init(unpackingAny:)
+            var any = Google_Protobuf_Any()
+            any.typeURL = _typeURL
+            any._storage._contentJSON = contentJSON
+            any._storage._valueData = nil
+            let m = try messageType.init(unpackingAny: any)
+            return try m.serializedData(partial: true)
+          } catch {
+            return Data()
+          }
+        }
+      }
+
+      return Data()
+    }
+    set {
+      _valueData = newValue
+      _message = nil
+      _contentJSON = nil
+    }
+  }
+
+  // The possible internal states for _value.
+  //
+  // Note: It might make sense to shift to using an enum for internal
+  // state instead to better enforce this; but that also means we could
+  // never got a model were we might also be able to cache the things
+  // we have.
+  var _valueData: Data? = Data()
+  var _message: Message?
+  var _contentJSON: Data?  // Any json parsed from with the @type removed.
+
+  init() {}
+
+  func copy() -> AnyMessageStorage {
```

Comment:
Bit of an aside:  Is there any benefit to using copy initializers instead of a copy method like this?  The method form needs to initialize every field before copying everything over; would a copy initializer avoid that?